### PR TITLE
Fix stale index deletion in snapshots for hashed prefix path type

### PR DESCRIPTION
**Pull request by ashking94**
### Description
If the index name has `.` period in it, then the stale index deletion fails as the period character interferes with the delimiter (which is also period character) by giving more parts than expected during the `String.split(".")` operation while interpreting the snapshot shards path file. The fix is to handle the variables after the index name as relative from last than from 0. This fixes the problem.

### Check List
- [x] Functionality includes testing.
- ~[ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.~
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -2549,25 +2561,32 @@ public void finalizeSnapshot(
...
-            deleteFromContainer(snapshotShardPathBlobContainer(), staleShardPaths);
-        } catch (IOException e) {
-            logger.warn(
-                new ParameterizedMessage(
-                    "Repository [{}] Exception during snapshot stale index deletion {}",
-                    metadata.name(),
-                    staleShardPaths
-                ),
-                e
-            );
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java:2570]: This change is only to put this block of code within a try/catch block.

File: server/src/main/java/org/opensearch/snapshots/SnapshotShardPaths.java [hunk 2]
```diff
@@ -92,18 +92,21 @@ public static SnapshotShardPaths fromXContent(XContentParser ignored) {
...
-        if (parts.length != 5) {
+        int len = parts.length;
+        if (len < 5) {
             throw new IllegalArgumentException("Invalid shard path format: " + shardPath);
         }
         try {
-            IndexId indexId = new IndexId(parts[1], getIndexId(parts[0]), Integer.parseInt(parts[3]));
-            int shardCount = Integer.parseInt(parts[2]);
+            // indexName is never used from within the ShardInfo returned
+            String indexName = len == 5 ? parts[1] : "";
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/snapshots/SnapshotShardPaths.java:107]: Get the index name for all cases.

File: server/src/main/java/org/opensearch/snapshots/SnapshotShardPaths.java [hunk 3]
```diff
@@ -103,8 +103,12 @@ public static ShardInfo parseShardPath(String shardPath) {
             throw new IllegalArgumentException("Invalid shard path format: " + shardPath);
         }
         try {
-            // indexName is never used from within the ShardInfo returned
-            String indexName = len == 5 ? parts[1] : "";
+            String indexName = shardPath.substring(
+                // First separator after index id
+                shardPath.indexOf(DELIMITER) + 1,
+                // Since we know there are exactly 3 fields at the end
+                shardPath.lastIndexOf(DELIMITER, shardPath.lastIndexOf(DELIMITER, shardPath.lastIndexOf(DELIMITER) - 1) - 1)
```
Comment (gbbafna) [inline server/src/main/java/org/opensearch/snapshots/SnapshotShardPaths.java:110]: can we simplify this 

```
    parts = shardPath.split('.')
    
    # The index_name is everything between the first and last two parts
    if len(parts) >= 5:
        index_name = '.'.join(parts[1:-3])
```

Comment (gbbafna) [reply server/src/main/java/org/opensearch/snapshots/SnapshotShardPaths.java:110]: Looks like this is simplest one .

**Review by gbbafna — APPROVED** [review]

**Pull request merged**